arch/arm/core/cortex_a_r/swap_helper.S

```diff
@@ -24,14 +24,71 @@
 _ASM_FILE_PROLOGUE
 
 GTEXT(z_arm_svc)
+#ifdef CONFIG_USE_SWITCH
+GTEXT(z_arm_context_switch)
+#else
 GTEXT(z_arm_do_swap)
+#endif /* CONFIG_USE_SWITCH */
 GTEXT(z_do_kernel_oops)
 #if defined(CONFIG_USERSPACE)
 GTEXT(z_arm_do_syscall)
 #endif
 
 GDATA(_kernel)
 
+#ifdef CONFIG_USE_SWITCH
+/*
+ * Routine to handle context switches
+ *
+ * This function is directly called either by _isr_wrapper() in case of
+ * preemption, or arch_switch() in case of cooperative switching.
+ *
+ * void z_arm_context_switch(struct k_thread *new, struct k_thread *old);
+ */
+SECTION_FUNC(TEXT, z_arm_context_switch)
+
+	ldr r2, =_thread_offset_to_callee_saved
+	add r2, r1, r2
+
+	stm r2, {r4-r11, sp, lr}
+
+	/* save old thread into switch handle which is required by
+	 * z_sched_switch_spin()
+	 */
+	str r1, [r1, #___thread_t_switch_handle_OFFSET]
```

Comment:
FWIW: this must be the **very last** thing done with the old thread context.  Since someone will be coming back here to add FPU support later, it might be nice for the comment to warn about that and/or mention the docs in arch_interface.h that explain the requirement.

Also, at least on SMP Cortex-A you need a write barrier immediately before this store, so that no other CPU can possibly see the earlier stores cross with it and miss some of the saved data.

Reply:
Done